src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs

```diff
@@ -27,42 +29,239 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-US/library/ms182269.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(NotCalledWithFinalizerRule, NotCalledRule, NotPassedThisRule, OutsideDisposeRule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            {
+                compilationContext.RegisterOperationBlockStartAction(operationBlockContext =>
+                {
+                    if (operationBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                    {
+                        return;
+                    }
+
+                    var method = (IMethodSymbol)operationBlockContext.OwningSymbol;
+                    if (method.IsExtern || method.IsAbstract)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new SuppressFinalizeAnalyzer(method, compilationContext.Compilation);
+
+                    operationBlockContext.RegisterOperationAction(analyzer.Analyze, OperationKind.InvocationExpression);
+
+                    operationBlockContext.RegisterOperationBlockEndAction(analyzer.OperationBlockEndAction);
+                });
+            });
+
+        }
+
+
+
+        private class SuppressFinalizeAnalyzer
+        {
+            private enum SuppressFinalizeUsage
+            {
+                CanCall,
+                MustCall,
+                MustNotCall
+            }
+
+            private bool _suppressFinalizeCalled = false;
+
+            private readonly Compilation _compilation;
+            private readonly IMethodSymbol _containingMethodSymbol;
+            private SemanticModel _semanticModel;
+            private IInvocationExpression _invocationExpression;
+            private IMethodSymbol _gcSuppressFinalizeMethodSymbol;
+            private bool _failFast;
+
+            public SuppressFinalizeAnalyzer(IMethodSymbol methodSymbol, Compilation compilation)
+            {
+                this._compilation = compilation;
+                this._containingMethodSymbol = methodSymbol;
+            }
+
+            public void Analyze(OperationAnalysisContext analysisContext)
+            {
+                _invocationExpression = (IInvocationExpression)analysisContext.Operation;
+                _gcSuppressFinalizeMethodSymbol = _compilation.GetTypeByMetadataName("System.GC").GetMembers("SuppressFinalize").OfType<IMethodSymbol>().SingleOrDefault();
```

Comment:
You should compute well known types/members in the CompilationStartAction, avoid registering any operation block action if gcSuppressFinalizeMethodSymbol == null, and if not store it as a readonly field of SuppressFinalizeAnalyzer.